### PR TITLE
Fix #1021: Implemented string deserializer that removes leading and trailing spaces

### LINKED ISSUE #1021: Configuration INI trailing spaces - regression
### Bug description
In current dev c3087aec93497f183a93f2bdfb9bbfec2fb56b00 setting a Coo with trailing spaces will result in an unclear exception:
```
Exception in thread "main" java.lang.IllegalArgumentException: Destination array is not large enough.
	at com.iota.iri.utils.Converter.trits(Converter.java:93)
	at com.iota.iri.model.HashFactory.create(HashFactory.java:48)
	at com.iota.iri.MilestoneTracker.<init>(MilestoneTracker.java:72)
	at com.iota.iri.Iota.<init>(Iota.java:56)
	at com.iota.iri.IRI$IRILauncher.main(IRI.java:69)
	at com.iota.iri.IRI.main(IRI.java:31)
```

This does not happen in v1.5.4 (latest release at the time) so it is either related to #969 or #724 

if related to Conf rework then it's worth looking at other places where this might happen in the INI file.
this can be set to wont-fix, but then it should be mentioned in the release notes as a known issue.

### Steps To Reproduce
1. set iota.iri to:
```
[IRI]
ZMQ_ENABLED = TRUE
TESTNET = TRUE
MWM = 9
SNAPSHOT_FILE = conf/snapshot.txt
COORDINATOR = NPCRMHDOMU9QHFFBKFCWFHFJNNQDRNDOGVPEVDVGWKHFUFEXLWJBHXDJFKQGYFRDZBQIFDSJMUCCQVICI 
MILESTONE_START_INDEX = 2
KEYS_IN_MILESTONE = 10
MAX_DEPTH = 1000
```
2. run IRI on that commit.
3. it will fail with the aformentioned Exception.

### Expected behaviour
Not sure, but I would expect trailing spaces to be trimmed.

### Actual behaviour
Unclear exit

### Errors
Posted above.

### PR DESCRIPTION
**Pull request by legacycode**
# Description

I added a unit test for ConfigFactory.java and implemented string deserializer that removes leading and trailing spaces.

Fix #1021 

- Enhancement (a non-breaking change which adds functionality)

# How Has This Been Tested?

- Run mvn clean install

# Checklist:

- [x] My code follows the style guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by legacycode** [comment]
Fixes #1021

File: src/main/java/com/iota/iri/conf/deserializers/CustomStringDeserializer.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+    /**
+     * Default constructor
+     */
+    public CustomStringDeserializer() {
+        super(String.class);
+    }
+
+    @Override
+    public String deserialize(JsonParser jsonParser, DeserializationContext deserializationContext) throws IOException, JsonProcessingException {
```
Comment (GalRogozinski) [inline src/main/java/com/iota/iri/conf/deserializers/CustomStringDeserializer.java]: `JsonProcessingException` extends `IOException`. I don't see a point in declaring them both...

File: src/test/java/com/iota/iri/conf/ConfigFactoryTest.java [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+import java.util.Properties;
+
+import static junit.framework.TestCase.assertTrue;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+
+/**
+ * Tests for the {@link ConfigFactory}
+ */
+public class ConfigFactoryTest {
```
Comment (GalRogozinski) [inline src/test/java/com/iota/iri/conf/ConfigFactoryTest.java]: Great! Note that there is also a related `ConfigTest` class

File: src/test/java/com/iota/iri/conf/ConfigFactoryTest.java [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+    @Test(expected = FileNotFoundException.class)
+    public void createFromFileTestnetWithFileNotFound() throws IOException {
+        File configFile = new File("doesNotExist.ini");
+        ConfigFactory.createFromFile(configFile, false);
+    }
+
+    private File createTestnetConfigFile(String testnet) throws IOException {
+        Properties properties = new Properties();
+        properties.setProperty("TESTNET", testnet);
+        properties.setProperty("ZMQ_ENABLED", " TRUE ");
```
Comment (GalRogozinski) [inline src/test/java/com/iota/iri/conf/ConfigFactoryTest.java]: Note that the output in the file contains `\` for values with leading whitespace:
`ZMQ_ENABLED=\ TRUE`
Did you mean for that?

File: src/test/java/com/iota/iri/conf/ConfigFactoryTest.java [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+        assertTrue("Expected iotaConfig as Testnet.", iotaConfig.isTestnet());
+    }
+
+    /**
+     * Creates and validates a Testnet {@link IotaConfig} with <code>TESTNET=true</code> in config file and
+     * <code>testnet: true</code> as method parameter for {@link ConfigFactory#createFromFile(File, boolean)}.
+     * @throws IOException when config file not found.
+     */
+    @Test
+    public void createFromFileTestnetWithTestnetTrueAndTrue() throws IOException {
```
Comment (GalRogozinski) [inline src/test/java/com/iota/iri/conf/ConfigFactoryTest.java]: Maybe you can delete https://github.com/iotaledger/iri/blob/dev/src/test/java/com/iota/iri/conf/ConfigTest.java#L82
and https://github.com/iotaledger/iri/blob/dev/src/test/java/com/iota/iri/conf/ConfigTest.java#L150

If you want I think using `AssertThat` gives nicer error messages. The way you did it is fine as well.

File: src/test/java/com/iota/iri/conf/ConfigFactoryTest.java [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+        assertTrue("Expected iotaConfig as instance of MainnetConfig.", iotaConfig instanceof MainnetConfig);
+        assertFalse("Expected iotaConfig as Mainnet.", iotaConfig.isTestnet());
+    }
+
+    /**
+     * Test if leading and trailing spaces are trimmed from string in properties file.
+     * @throws IOException when config file not found.
+     */
+    @Test
+    public void createFromFileTestnetWithTrailingSpaces() throws IOException {
```
Comment (GalRogozinski) [inline src/test/java/com/iota/iri/conf/ConfigFactoryTest.java]: Perhaps we should a nice separation where `ConfigTest` tests parsing and `ConfigFactoryTest` tests creation?

**Comment by jakubcech** [comment]
Have a good vacation! :)

**Comment by GalRogozinski** [comment]
@legacycode 
I have cherry picked your commits and merged them to the `release-v1.5.5` branch in #1050 since we want to release soon and have this bug fixed.

It will be great if you could perform a PR on `dev` that fixes my comments after you get back from your vacation.

I wish you an enjoyable vacation.

**Pull request closed without merging**